node_package/src/serverRenderReactComponent.ts

```diff
@@ -6,154 +6,180 @@ import createReactOutput from './createReactOutput';
 calls renderToString, that takes one parameter. You need to add an extra unused parameter to identify this function
 as a renderFunction and not a simple React Function Component.`);
-        throw error;
-      }
-    };
-
-    if (isServerRenderHash(reactRenderingResult)) {
-      renderResult = processServerRenderHash();
-    } else if (isPromise(reactRenderingResult)) {
-      renderResult = processPromise() as Promise<string>;
-    } else {
-      renderResult = processReactElement();
-    }
-  } catch (e: any) {
-    if (throwJsErrors) {
-      throw e;
-    }
-
-    hasErrors = true;
-    renderResult = handleError({
-      e,
-      name,
-      serverSide: true,
-    });
-    renderingError = e;
+    throw error;
   }
+}
 
-  const consoleReplayScript = buildConsoleReplay();
-  const addRenderingErrors = (resultObject: RenderResult, renderError: RenderingError) => {
-    // Do not use `resultObject.renderingError = renderError` because JSON.stringify will turn it into '{}'.
-    resultObject.renderingError = { // eslint-disable-line no-param-reassign
-      message: renderError.message,
-      stack: renderError.stack,
-    };
-  };
+function processRenderingResult(result: CreateReactOutputResult, options: RenderOptions): RenderState {
+  if (isServerRenderHash(result)) {
+    return processServerRenderHash(result, options);
+  }
+  if (isPromise(result)) {
+    return { result: processPromise(result, options.renderingReturnsPromises), hasErrors: false };
+  }
+  return { result: processReactElement(result), hasErrors: false };
+}
 
-  if (renderingReturnsPromises) {
-    const resolveRenderResult = async () => {
-      let promiseResult: RenderResult;
-
-      try {
-        promiseResult = {
-          html: await renderResult,
-          consoleReplayScript,
-          hasErrors,
-        };
-      } catch (e: any) {
-        if (throwJsErrors) {
-          throw e;
-        }
-        promiseResult = {
-          html: handleError({
-            e,
-            name,
-            serverSide: true,
-          }),
-          consoleReplayScript,
-          hasErrors: true,
-        };
-        renderingError = e;
-      }
-
-      if (renderingError !== null) {
-        addRenderingErrors(promiseResult, renderingError);
-      }
-
-      return promiseResult;
-    };
-
-    return resolveRenderResult();
+function handleRenderingError(e: unknown, options: { componentName: string, throwJsErrors: boolean }) {
+  if (options.throwJsErrors) {
+    throw e;
   }
+  const error = e instanceof Error ? e : new Error(String(e));
+  return {
+    hasErrors: true,
+    result: handleError({ e: error, name: options.componentName, serverSide: true }),
+    error,
+  };
+}
 
-  const result: RenderResult = {
-    html: renderResult as string,
+function createResultObject(html: string | null, consoleReplayScript: string, hasErrors: boolean, error?: RenderingError): RenderResult {
```

Comment:
I thought of replacing `html`, `hasErrors` and `error` arguments with one argument of type `RenderState`, but I found that the `result` in `RenderState` can be a string or promise not just a string and it would take more casts and checks in other functions, so I prefer to keep this style.